### DISCUSSION TITLE
Crazyflie with custom firmware fails to land after tracking velocity setpoints

### DISCUSSION THREAD

**Discussion by girayus** [bitcraze/discussions#744] [Q&A] [answered]
Hello there!

I'm testing my navigation algorithm onboard via the app layer. I'm using Crazyswarm only to send position data to Crazyflie and change the value of my custom mission control parameter 'flightstate'. Everything goes perfectly up to landing. However, when the mission completes, Crazyflie keeps tracking the last setpoint command for a short time, cuts power from the motors, and falls vertically on the ground. I read in forums that switching from low-level to high-level commander is no longer required, and related functions are deprecated. Using debug print, I checked that Crazyswarm successfully changes the value of the 'flightstate' parameter for landing. But Crazyflie cannot execute high-level land command after tracking velocity setpoints. I removed the velocity calculation part of my code and added it below.

My question is, what should I change/add in my code to land the Crazyflie successfully?
<pre><code>
#include "string.h"
#include "stdint.h"
#include "stdbool.h"
#include "app.h"
#include "FreeRTOS.h"
#include "task.h"
#include "estimator_kalman.h"
#include "debug.h"
#include "log.h"
#include "param.h"
#include "radiolink.h"
#include "configblock.h"
#include "crtp_commander_high_level.h"
#include "commander.h"
#include "pptraj.h"
#include "usec_time.h"
#include "peer_localization.h"
#include "math3d.h"
#define DEBUG_MODULE "P2P"
#define dt_msec 100
static paramVarId_t paramIdResetKalman;
static paramVarId_t paramIdFlightState;

static void setHoverSetpoint(setpoint_t *setpoint, float vx, float vy, float vz)
{
  float yawRate = 0.0;
  setpoint->mode.yaw = modeVelocity;
  setpoint->attitudeRate.yaw = yawRate;
  setpoint->mode.x = modeVelocity;
  setpoint->mode.y = modeVelocity;
  setpoint->mode.z = modeVelocity;
  setpoint->velocity.x = vx;
  setpoint->velocity.y = vy;
  setpoint->velocity.z = vz;
  setpoint->velocity_body = false;
}

void appMain() {
  //DEBUG_PRINT("Waiting for activation ...\n");
  static setpoint_t setpoint;
  static uint8_t flightstate;
  flightstate = 0;
  PARAM_GROUP_START(fmodes)
  PARAM_ADD_CORE(PARAM_UINT8, ufs, &flightstate)
  PARAM_GROUP_STOP(fmodes)
  paramIdFlightState = paramGetVarId("fmodes", "ufs");
  logVarId_t idX = logGetVarId("stateEstimate", "x");
  logVarId_t idY = logGetVarId("stateEstimate", "y");
  logVarId_t idZ = logGetVarId("stateEstimate", "z");
  paramSetInt(paramIdResetKalman, 1);
  while(1) {
    //idle state
    if (flightstate == 0){
      flightstate = paramGetInt(paramIdFlightState);
      vTaskDelay(M2T(100));
    }
    //take off
    if (flightstate == 1){
      vTaskDelay(M2T(1000));
      crtpCommanderHighLevelTakeoff(take_off_height, 2.5);
      vTaskDelay(M2T(4000));
      flightstate = 2;
    }
    //check whether take off is completed
    if ((flightstate == 2) && crtpCommanderHighLevelIsTrajectoryFinished()){
        flightstate = 3;
    }
    //flocking
    if (flightstate == 3){
      //.........
      //--Code calculates vx, vy and vz--//
      //.........
      //set control inputs
      setHoverSetpoint(&setpoint, vx, vy, vz);
      commanderSetSetpoint(&setpoint, 3);
      vTaskDelay(M2T(dt_msec)); //sleep for dt, assuming computation time is negligible
      flightstate = paramGetInt(paramIdFlightState); //check whether state is set to land or not
    }
    //land
    if (flightstate == 4){
      crtpCommanderHighLevelLand(0.03, 3.0); 
      vTaskDelay(M2T(4000));
      return;
    }
  }
}
</code></pre>

**Comment by krichardsson** [comment, marked as answer]
I'm not exactly sure what is happening, but my best guess is that it is related to the setpoint priority feature. I can see that you are using level 3 and I think this will block the high level commander setpoints (with priority 1) when you want to land. Try calling [`commanderRelaxPriority()`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/commander.c#L94) before landing to reset the priority.

**Reply by girayus** [reply]
I called [commanderRelaxPriority()](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/commander.c#L94) just before the landing command and Crazyflie successfully landed. Thanks a lot for your help!
